docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,236 @@
+- New seeds became available in a region closer to/in the region of the workers and the control plane should be moved there to improve latency
+- Gardener ring, which is a self-supporting setup/underlay for a highly available (usually cross-region) Gardener deployment
+
+
+## Goals
+
+* Provide a mechanism to migrate the control plane of a shoot cluster from one seed to another
+* The mechanism should support migration from a seed which is no longer reachable (Disaster Recovery)
+* The mechanism can automatically decide whether a seed is unhealthy and control planes need to be migrated to a new seed.
+* The mechanism needs to take into account that gardenlet will be used for shoot reconciliation very soon
+* The shoot cluster nodes are preserved and continue to run the workload, but will talk to the new control plane after the migration completes
+* Extension controllers implement a mechanism which allows them to store their state or to be restored from an already existing state on a different seed cluster.
+* The already existing shoot reconciliation flow is reused for migration with minimum changes
+
+
+## Terminology
+
+__Source Seed__ is the seed which currently hosts the control plane of a Shoot Cluster
+
+__Destination Seed__ is the seed to which the control plane is being migrated
+
+
+## Resources and controller state which have to be migrated between two seeds:
+
+**Note:** The following lists are just FYI and are meant to show the current resources which need to be moved to the __Destination Seed__
+
+### Secrets
+
+Gardener has preconfigured lists of needed secrets which are generated when a shoot is created and deployed in the seed. Following is a minimum set of secrets which must be migrated to the New Seed. Other secrets can be regenerated from them.
+
+- ca
+- ca-front-proxy
+- static-token
+- ca-kubelet
+- ca-metrics-server
+- etcd-encryption-secret
+- kube-aggregator
+- kube-apiserver-basic-auth
+- kube-apiserver
+- service-account-key
+- ssh-keypair
+
+### Custom Resources and state of extension controllers
+
+Gardener deploys custom resources to the Host Seed cluster which are reconciled by extension controllers. The state of these controllers and any additional resources they create is independent of gardener and must also be migrated to the New Seed. Following is a list of custom resources, and the state which is generated by them that has to be migrated.
+
+- **BackupBucket**: nothing relevant for migration
+- **BackupEntry**: nothing relevant for migration
+- **ControlPlane**: nothing relevant for migration
+- **DNSProvider**/DNSEntry: nothing relevant for migration
+- **Extensions**: migration of state needs to be handled individually
+- **Infrastructure**: terraform state
+- **Network**: nothing relevant for migration
+- **OperatingSystemConfig**: nothing relevant for migration
+- **Worker**: Machine-Controller-Manager related objects: machineclasses, machinedeployments, machinesets, machines
```

Comment:
Did you meanwhile talk to @hardikdr and @prashanth26 wrt. to whether it would be an option to make MCM capable of recreating the `Machine` objects on its own based on existing `Node` objects? What's their opinion?

Reply:
I asked @amshuman-kr but still haven't pinged him if there is any new information on the topic

Reply:
Well..., can't we just simply schedule a meeting with the guys now? 15-30min are certainly enough. We don't need to talk about the details, it's just important to get to know their opinion on the approach and whether it's worth it to follow it or to handle the machine object migration solely in the `Worker` controller.

Reply:
I think re-creating machine objects from node objects should possible. We might have to bake in logic to re-create machines using some metadata stored on VM objects (using tags maybe). However, the current MCM doesn't support this feature. I think it should be possible in the future but might take time. So these are the options I can think of. 

Option-1: If the shoot migration topic is planned to be completed soon I would suggest to take a backup of MCM objects for now and restore them. Later, we can implement this logic in MCM and get rid of this part. We can create an issue to track the same. 

Option-2: If there is ample time we could go with the assumption that MCM would re-create machines/machine-sets/machine-deployments from node objects. Just to keep in mind would be that it will require some time for us to validate the same and if it's not possible we might have to fall back to option-1. 

We can discuss more regarding this offline later today.